src/libcore/array/mod.rs

```diff
@@ -387,6 +387,74 @@ where
     }
 }
 
+/// Non-zero numbers
+///
+/// We require following from this trait:
+///  - [(); 0] does **not** implement NonZero.
+///  - [(); N] where N != 0 **implements** NonZero.
+///  As side effect of selected approach, this trait is also implemented
+///  for all other types, but it does not matter.
+#[unstable(feature = "array_default_internals", reason = "implementation detail", issue = "none")]
+pub auto trait NonZero {}
+#[unstable(feature = "array_default_internals", reason = "implementation detail", issue = "none")]
+impl !NonZero for [(); 0] {}
+
+// This is implementation for all non-zero-sized arrays
+#[stable(since = "1.4.0", feature = "array_default")]
+impl<T: Default, const N: usize> Default for [T; N]
+where
+    [(); N]: NonZero,
+    [T; N]: LengthAtMost32,
+{
+    #[inline]
+    fn default() -> [T; N] {
+        use crate::mem::MaybeUninit;
+        let mut data: MaybeUninit<[T; N]> = MaybeUninit::uninit();
+        // invariant: first `init` items are initialized
+        struct Guard<T, const N: usize> {
+            data: *mut T,
+            init: usize,
+        }
```

Comment:
This is effectively implementing `FromIterator` for arrays. I think extracting this making it reusable across other places would make sense.

Reply:
Well, I imagine something like:
```rust
// in core::mem
pub(crate) struct PartialBuffer<T, const N> {
    buf: [MaybeUninit<T>; N],
    init: usize
}
impl PartialBuffer<_, _> {

    pub(crate) fn new() -> Self {..}

     pub(crate) unsafe fn push(&mut self, value: T) {..}
     // or maybe pub(crate) unsafe fn next_place(&mut self) -> &mut MaybeUninit<T> {..} to reduce copying

     pub(crate) unsafe fn assume_init(self) -> [T; N]
     pub(crate) fn get_ref(&self) -> &[T] {}
     pub(crate) fn get_mut(&mut self) -> &mut [T]  
}


```